File: Problems/1043/main.go
package main

import "slices"

func main() {}

func maxSumAfterPartitioning(arr []int, k int) int {
	// Naive Solution, Partiton all Possibile Solution
	//
	// DP to the rescue, add or not to add. Found by thinking about merge sort.
	// Time Complexity: O(nk), Space Complexity: O(n)
	res := make([]int, len(arr)+1)
	m := len(arr)

	res[m] = 0
	res[m-1] = arr[m-1]

	for i := m - 2; i >= 0; i-- {
		best := 0
		maxInK := -1
		j := i
		for _k := 0; _k < k; _k++ {
			j++
			if j > m {
				break
			}
			if arr[j-1] > maxInK {
				maxInK = arr[j-1]
			}
			tmp := maxInK*(_k+1) + res[j]
			if tmp > best {
				best = tmp
			}
		}
		res[i] = best
	}

	return res[0]
}

func maxSumAfterPartitioningReadableVersion(arr []int, k int) int {
	// Naive Solution, Partiton all Possibile Solution
	//
	// DP to the rescue, add or not to add. Found by thinking about merge sort.
	// Time Complexity: O(nk), Space Complexity: O(n)
	res := make([]int, len(arr)+1)
	m := len(arr)

	getMaxSum := func(arr []int) int {
		return slices.Max(arr) * len(arr)
	}
	res[m] = 0
	res[m-1] = arr[m-1]

	for i := m - 2; i >= 0; i-- {
		best := 0
		for j := 1; j <= k && i+j <= m; j++ {
			tmp := getMaxSum(arr[i:i+j]) + res[i+j]
			if tmp > best {
				best = tmp
			}
		}
		res[i] = best
	}

	return res[0]
}

File: Problems/1043/main.cpp
#include <vector>
#include <iostream>
#include <string>
#include <map>
#include <stack>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>

using namespace std;

class Solution
{
public:
    int maxSumAfterPartitioning(vector<int> &arr, int k)
    {
        int n = arr.size();
        vector<int> dp(n + 1, 0);
        // dp[i]: maximum value of arr[0:i] after partition

        for (int i = 1; i <= n; i++)
        {
            int MAX = -1;
            for (int j = i; j >= max(1, i - k + 1); j--)
            {
                MAX = max(MAX, arr[j - 1]);
                dp[i] = max(dp[i], MAX * (i - j + 1) + dp[j - 1]);
            }
        }

        return dp[n];
    }
};
